### ISSUE TITLE
【论文复现】UGATIT 方差 torch.var相当的方法，paddle是否有提供

### ISSUE THREAD

**Issue by sincerevertigo** [PaddlePaddle/Paddle#26109]
如题。
paddle1.8
谢谢。

**Comment by oytang** [comment]
1.8没有方差算子，需要自己写。一定想直接用方差算子的话，2.0-beta里有。需要注意`torch.var`计算时考虑了一个自由度，因此通过paddle的算子来直接组合得到有些困难。我的方法是先转成`ndarray`使用`numpy`里面的`numpy.var`计算方差再转回paddle。

**Comment by sincerevertigo** [comment]
@Chemino 

非常感谢！！！
按照上面的思路，写了一下代码。


input_numpy = np.array(input)
input_var = np.var(input_numpy, axis=[2, 3], keepdims=True)
input_tensor = fluid.Tensor()
input_tensor.set(input_numpy, fluid.CPUPlace())


上面这个方法可以吗，由于时间原因，还没有做最终的确认。

**Comment by oytang** [comment]
@sincerevertigo 

大体思路没有问题，但我常用的是paddle的动态图，所以转入和转出我分别用的是Tensor.numpy()方法和dygraph.to_variable(x)算子，不确定你的是否可行。此外，计算方差时要考虑自由度ddof，这是复现pytorch算法的关键。

**Comment by sincerevertigo** [comment]
@Chemino 
多谢回答。

在forword函数里，是不是就不能用图的方法了？
Tensor.numpy()
这个方法是直接点出来的？
input_numpy = input.numpy()
>计算方差时要考虑自由度ddof，这是复现pytorch算法的关键。
现在还没看出来源码是总体方差还是样本方差，再研究研究。

**Comment by sandyhouse** [comment]
应该是可以使用.numpy方法的，你在这里使用的目的是什么。

**Comment by sincerevertigo** [comment]
@sandyhouse @Chemino 

不好意思，问题形容的不太好。
我重新整理了一下。

class adaILN(fluid.dygraph.Layer):
    def __init__(self, num_features, eps=1e-5):
        super(adaILN, self).__init__()
        self.eps = eps
        self.rho = Parameter(torch.Tensor(1, num_features, 1, 1))
        self.rho.data.fill_(0.9)
    def forward(self, input, gamma, beta):
        、、、
        in_mean, in_var = torch.mean(input, dim=[2, 3], keepdim=True), torch.var(input, dim=[2, 3], keepdim=True)
        、、、


具体来说就是要把上面的代码进行paddle的移植
torch.mean
用的
fluid.layers.reduce_mean

torch.var
现在不知道该怎么办。

谢谢。

**Comment by sandyhouse** [comment]
求方差的话直接按照方差的定义计算是不是就可以了。
![image](https://user-images.githubusercontent.com/6737864/89892448-3ded4e00-dc09-11ea-8960-2cd00adb3554.png)

例如，要求一个二维矩阵A每行的方差，可以按照下面的步骤：
1. 求每行的均值，得到二维矩阵B
2. C = elementwise_sub(A, B)得到每行中每个元素和其改行均值的差值
3. D = square(C)
4. E = elementwise_div(D, A.shape[1])
5. 用reduce_sum计算最后的方差值。

**Comment by sincerevertigo** [comment]
@Chemino 
使用了2.0的函数
paddle.Tensor.var(input, axis=[2, 3], keepdim=True)
这样的话，似乎可以直接用。
但是，程序现在还跑不通，没有验证，等验证后，更新。
十分感谢！！！

**Comment by liuxianyi** [comment]
> @Chemino
> 使用了2.0的函数
> paddle.Tensor.var(input, axis=[2, 3], keepdim=True)
> 这样的话，似乎可以直接用。
> 但是，程序现在还跑不通，没有验证，等验证后，更新。
> 十分感谢！！！

2.0好像gpu版本不匹配，你可以跑通吗？

**Comment by liuxianyi** [comment]
![image](https://user-images.githubusercontent.com/33951338/90085009-78f69b00-dd49-11ea-95da-56fa1b43dad1.png)
@ @sandyhouse 这是我按照你的方法写的，但是会出现w权值不能传播的情况？？求解？
![image](https://user-images.githubusercontent.com/33951338/90085068-a6434900-dd49-11ea-98a8-9a9183b6a1c7.png)

**Comment by sandyhouse** [comment]
计算in_mean1 = fluid.layers.reduce_mean这个地方，keep_dim是不是应该为True @liuxianyi

**Comment by liuxianyi** [comment]
> 计算in_mean1 = fluid.layers.reduce_mean这个地方，keep_dim是不是应该为True @liuxianyi

设置为Ture还是会出现w无法传播的错误信息啊

**Comment by sandyhouse** [comment]
给个完整的可复现代码吧

**Comment by liuxianyi** [comment]
> 给个完整的可复现代码吧
https://github.com/PaddlePaddle/Paddle/issues/25141#issue-642053156
我的情况和这个大哥一样，你们给的回答是在1.8.2bug修复了，但是现在使用不了1.8.2,会报错cudnn需要升级，求解答？，谢谢！

**Comment by oytang** [comment]
我根据Paddle的2.0.0a版本中`paddle.tensor.var`的原代码改编了以下代码，这样就可以在1.8中完成实现，并且相比于转numpy进行计算而言，使用paddle原生计算应该能带来效率上的优势。代码如下：
```
def var(input, dim=None, keep_dim=False, unbiased=True, name=None):
    rank = len(input.shape)
    dims = dim if dim != None and dim != [] else range(rank)
    dims = [e if e >= 0 else e + rank for e in dims]
    inp_shape = input.shape
    mean = layers.reduce_mean(input, dim=dim, keep_dim=True, name=name)
    tmp = layers.reduce_mean((input - mean)**2, dim=dim, keep_dim=keep_dim, name=name)
    if unbiased:
        n = 1
        for i in dims:
            n *= inp_shape[i]
        factor = n / (n - 1.0) if n > 1.0 else 0.0
        tmp *= factor
    return tmp
```
